### DISCUSSION TITLE
How to render a zone in decoupled CMS using Razor Pages

### DISCUSSION THREAD

**Discussion by dodyg** [OrchardCMS/OrchardCore#11753] [Q&A]
I have checked this issue https://github.com/OrchardCMS/OrchardCore/issues/6839 but it's not clear what's the working method is. 

```
For RazorPages you inherit from OrchardCore.DisplayManagement.RazorPages.Page
and For MVC Views rendered from Controller inherit from OrchardCore.DisplayManagement.Razor.RazorPage or OrchardCore.DisplayManagement.Razor.RazorPage<TModel>
```

This recommendation doesn't work.

**Comment by dodyg** [comment]
There is a similar question here that is not resolved https://github.com/OrchardCMS/OrchardCore/discussions/10606.

**Comment by ns8482e** [comment]
To use OrchardCore display management `theming` and `zones`  you should not set `_Layout` directly in your razor pages,  instead you use `Layout.cshtml` shape and your `csthml` should inherit from either of following.

For Razor pages, Place  in `/Pages` folder and inherit from `OrchardCore.DisplayManagement.RazorPages.Page`
For Views ( MVC and Shapes) in `/Views` folder and inherit from  `OrchardCore.DisplayManagement.Razor.RazorPage`

Btw there is working sample in `OrchardCore.Demo`

**Reply by dodyg** [reply]
> To use OrchardCore display management theming and zones you should not set _Layout directly in your razor pages, instead you use Layout.cshtml shape and your csthml should inherit from either of following.

Maybe I am dumb but I really don't understand how this would look like. 

I have problem figuring out what's going on here https://github.com/OrchardCMS/OrchardCore/tree/main/src/OrchardCore.Modules/OrchardCore.Demo because it is a mismatch between MVC views and Razor Pages.

I have a very simple Razor Pages only website and it has one `Pages/Shared/_Layout`.

**Reply by dodyg** [reply]
![scope](https://user-images.githubusercontent.com/1761249/169832369-3a8912d0-59fd-42e7-988b-6cd03323b567.png)

So this is just a simple decoupled CMS. I've tried several variations of putting `Layout.cshtml` in the `Views` folder and it either breaks the app or simply does not display layout.

**Reply by dodyg** [reply]
> For Razor pages, Place in /Pages folder and inherit from OrchardCore.DisplayManagement.RazorPages.Page
>For Views ( MVC and Shapes) in /Views folder and inherit from OrchardCore.DisplayManagement.Razor.RazorPage

In this configuration

![layout-1](https://user-images.githubusercontent.com/1761249/169833622-48880b3f-46ba-40c9-acc9-75378acb91b7.png)

![layout-2](https://user-images.githubusercontent.com/1761249/169834351-4be0c0a5-92bd-4c46-815c-36efb26c959d.png)

The result is the content rendered without the layout

![theme-1](https://user-images.githubusercontent.com/1761249/169833950-9fe5ea39-5ae7-4cfd-8557-b90f17260bdf.png)

**Reply by ns8482e** [reply]
Get rid of `_ViewStart.cshtml` 

To use theming
 - You don't define your `_Layout` in asp.net core to enable orchard display management - so there is not need to define `_ViewStart.cshtml` 
 - As orchard display management adds asp.net core layout for you - so you define your layout using shape  i.e. `Views\Layout.cshtml`

**Reply by dodyg** [reply]
The `_ViewStart.cshtml` is already empty.

![empty](https://user-images.githubusercontent.com/1761249/169965340-cb90a77e-b4a8-4f95-9271-5f4e9808e282.png)


I wonder if I a missing something else. 

![program-cs](https://user-images.githubusercontent.com/1761249/169965529-105afba1-5ecf-401f-8f22-44b6656b61a9.png)

**Reply by dodyg** [reply]
The funny thing is that with above configuration, I lose the admin layout

![admin](https://user-images.githubusercontent.com/1761249/169966379-d0a6589b-affd-4aed-bb04-058c9349424f.png)

**Reply by ns8482e** [reply]
You don’t need _ViewStart.cshtml even empty one- you need to delete it

**Reply by dodyg** [reply]
I've deleted them. The result is what I posted above.

**Reply by ns8482e** [reply]
The caveat with theme and display management is that Host shape is always resolved  meaning "Layout.cshtml" in host web is resolved by both front-end theme and admin theme.

**Reply by ns8482e** [reply]
@dodyg  you may want to rename `Layout.cshtml` to `Layout-Frontend.cshtml` 

and on each  razor `Page` on front-end  set `ViewLayout = "Layout__Frontend" ; `

OR you can create theme and create "Layout.cshtml" there and make that theme active from admin

**Reply by dodyg** [reply]
It generates error when I use `_ViewStart.cshtml`.

![error](https://user-images.githubusercontent.com/1761249/170091145-514be764-73fa-4aff-bcd7-990c86d5bf0e.png)

**Reply by dodyg** [reply]
When I use it directly on the page, it doesn't pick it up.

![directly](https://user-images.githubusercontent.com/1761249/170091475-5882bfff-d492-4cc2-bcac-08de448d5884.png)

**Reply by Skrypt** [reply]
Try with :  

```C#
@{
    Layout = "something";
}
```

**Reply by ns8482e** [reply]
No _viewstart.cshtml

ViewLayout should work - it's working in below screenshot, notice that the table border is coming from Layout-Frontend.cshtml

## /Pages/TestPage.cshtml
![image](https://user-images.githubusercontent.com/23270244/170095263-34cdab70-87b9-4114-803b-b7c8170c61b2.png)

## /Views/Layout-Frontend.cshtml
![image](https://user-images.githubusercontent.com/23270244/170095809-3399103e-95af-4c82-bae0-705648ed6afb.png)

## output
![image](https://user-images.githubusercontent.com/23270244/170096106-ada56ab9-01b4-4dfc-a692-b0208bb8bf47.png)

**Reply by ns8482e** [reply]
Also `@section` doesn't work with display management - use `<zone>` instead

**Reply by dodyg** [reply]
Unfortunately I still can't replicate your result. I've created a minimalistic project here https://github.com/dodyg/practical-aspnetcore/tree/net6.0/projects/orchard-core/decoupled-cms.

There must be something that I am missing. But there is barely any code now.

![wrong](https://user-images.githubusercontent.com/1761249/170225004-94d3bc59-08fe-4c92-8cb6-12d03333eb07.png)
![wrong-2](https://user-images.githubusercontent.com/1761249/170225016-be81e870-c18b-453b-b12d-6d27ec1069f3.png)

**Reply by ns8482e** [reply]
It's working for me,  may be you clean and try again.  The only change I made is removed PreserveCompilationReferences from csproj.  net 6 doesn't support runtime compilation. use dotnet watch hot reload instead.

**Reply by dodyg** [reply]
Unfortunately removing `PreserveCompilationReferences ` and cleaning and rebuilding makes no difference. It's the same result with `dotnet watch` or `dotnet run`.

**Reply by dodyg** [reply]
Changing the layout to liquid also doesn't work

![liquid](https://user-images.githubusercontent.com/1761249/170310848-8d47bee5-dc61-4e9a-8da6-371c2f986715.png)

**Reply by dodyg** [reply]
The funny thing is that I've implemented a bunch of decoupled CMS websites and never considered using Zone and Layers features. So it's possible to live without them. Now this is more like an intellectual curiosity why it doesn't work in my case.

**Reply by ns8482e** [reply]
Try removing  nuget config on parent directory.

**Reply by ns8482e** [reply]
I feel you might have some incorrect package references

**Reply by dodyg** [reply]
I did this and it didn't help. I am going to experiment some more and see if I can make this work.
